### DISCUSSION TITLE
shillelagh error: unable to open database file when using shillelagh

### DISCUSSION THREAD

**Discussion by zinginging** [apache/superset#23717] [Q&A / Help]
using shillelagh in docker demo from apache/superset:2.1.0

install shillelagh dep in `superset.Dockerfile`

```dockerfile
FROM apache/superset:2.1.0
USER root
RUN pip install 'shillelagh[all]'
USER superset
```

```shell
# build image
docker build . -t superset:0.1test -f ./superset.Dockerfile

# run 
docker run -d -p 8088:8088 -e "SUPERSET_SECRET_KEY=123456" --name superset superset:0.1test

# Setup local admin account
docker exec -it superset superset fab create-admin \
              --username admin \
              --firstname Superset \
              --lastname Admin \
              --email admin@superset.com \
              --password admin

# Migrate local DB to latest
docker exec -it superset superset db upgrade

# Setup roles
docker exec -it superset superset init
```

config shillelagh in superset 

```shell
display name: Shillelagh
sqlallchemy uri: shillelagh+safe://

Advanced -> Other -> Engine Parameters
{
  "connect_args": {
    "adapters": [
      "datasetteapi",
      "genericjsonapi",
      "githubapi",
      "htmltableapi",
      "s3selectapi",
      "socrataapi",
      "weatherapi"
    ]
  }
}
```

run sql in SQLlab 
```sql
SELECT
 date, administered_dose1_recip_4
FROM "https://data.cdc.gov/resource/unsk-b7fc.json"
WHERE location = 'US'
ORDER BY date DESC
LIMIT 10;
```

Followed the doc https://preset.io/blog/accessing-apis-with-superset/ from @betodealmeida get errors as below. Have you ever had a similar problem as beloow? Thanks a lot.

SupersetErrorsException
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1517, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1503, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/usr/local/lib/python3.8/site-packages/flask_appbuilder/security/decorators.py", line 95, in wraps
    return f(self, *args, **kwargs)
  File "/app/superset/views/base_api.py", line 122, in wraps
    raise ex
  File "/app/superset/views/base_api.py", line 113, in wraps
    duration, response = time_function(f, self, *args, **kwargs)
  File "/app/superset/utils/core.py", line 1586, in time_function
    response = func(*args, **kwargs)
  File "/app/superset/views/base_api.py", line 85, in wraps
    return f(self, *args, **kwargs)
  File "/app/superset/utils/log.py", line 266, in wrapper
    value = f(*args, **kwargs)
  File "/app/superset/sqllab/api.py", line 256, in execute_sql_query
    command_result: CommandResult = command.run()
  File "/app/superset/sqllab/commands/execute.py", line 120, in run
    raise ex
  File "/app/superset/sqllab/commands/execute.py", line 102, in run
    status = self._run_sql_json_exec_from_scratch()
  File "/app/superset/sqllab/commands/execute.py", line 157, in _run_sql_json_exec_from_scratch
    raise ex
  File "/app/superset/sqllab/commands/execute.py", line 152, in _run_sql_json_exec_from_scratch
    return self._sql_json_executor.execute(
  File "/app/superset/sqllab/sql_json_executer.py", line 111, in execute
    raise SupersetErrorsException(
superset.exceptions.SupersetErrorsException: [SupersetError(message='shillelagh error: unable to open database file', error_type=<SupersetErrorType.GENERIC_DB_ENGINE_ERROR: 'GENERIC_DB_ENGINE_ERROR'>, level=<ErrorLevel.ERROR: 'error'>, extra={'engine_name': 'Shillelagh', 'issue_codes': [{'code': 1002, 'message': 'Issue 1002 - The database returned an unexpected error.'}]})]
2023-04-17 12:19:46,273:WARNING:superset.views.base:SupersetErrorsException
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1517, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1503, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/usr/local/lib/python3.8/site-packages/flask_appbuilder/security/decorators.py", line 95, in wraps
    return f(self, *args, **kwargs)
  File "/app/superset/views/base_api.py", line 122, in wraps
    raise ex
  File "/app/superset/views/base_api.py", line 113, in wraps
    duration, response = time_function(f, self, *args, **kwargs)
  File "/app/superset/utils/core.py", line 1586, in time_function
    response = func(*args, **kwargs)
  File "/app/superset/views/base_api.py", line 85, in wraps
    return f(self, *args, **kwargs)
  File "/app/superset/utils/log.py", line 266, in wrapper
    value = f(*args, **kwargs)
  File "/app/superset/sqllab/api.py", line 256, in execute_sql_query
    command_result: CommandResult = command.run()
  File "/app/superset/sqllab/commands/execute.py", line 120, in run
    raise ex
  File "/app/superset/sqllab/commands/execute.py", line 102, in run
    status = self._run_sql_json_exec_from_scratch()
  File "/app/superset/sqllab/commands/execute.py", line 157, in _run_sql_json_exec_from_scratch
    raise ex
  File "/app/superset/sqllab/commands/execute.py", line 152, in _run_sql_json_exec_from_scratch
    return self._sql_json_executor.execute(
  File "/app/superset/sqllab/sql_json_executer.py", line 111, in execute
    raise SupersetErrorsException(
superset.exceptions.SupersetErrorsException: [SupersetError(message='shillelagh error: unable to open database file', error_type=<SupersetErrorType.GENERIC_DB_ENGINE_ERROR: 'GENERIC_DB_ENGINE_ERROR'>, level=<ErrorLevel.ERROR: 'error'>, extra={'engine_name': 'Shillelagh', 'issue_codes': [{'code': 1002, 'message': 'Issue 1002 - The database returned an unexpected error.'}]})]

**Comment by betodealmeida** [comment]
It looks like the `administered_dose1_recip_4` is no longer there. This query works for me:

```sql
SELECT
*
FROM "https://data.cdc.gov/resource/unsk-b7fc.json"
WHERE location = 'US'
ORDER BY date DESC
LIMIT 10;
```

**Reply by zinginging** [reply]
What's your superset and shillelagh version?

**Reply by betodealmeida** [reply]
I'm running both from master/main. Is the query above not working for you?

**Reply by zinginging** [reply]
Same error occurs  using the `select * xx` query. 

[SupersetError(message='shillelagh error: unable to open database file', error_type=<SupersetErrorType.GENERIC_DB_ENGINE_ERROR: 'GENERIC_DB_ENGINE_ERROR'>, level=<ErrorLevel.ERROR: 'error'>, extra={'engine_name': 'Shillelagh', 'issue_codes': [{'code': 1002, 'message': 'Issue 1002 - The database returned an unexpected error.'}]})]

**Reply by betodealmeida** [reply]
Huh, weird, let me give a try with Superset 2.1.0 and the latest shillelagh.

**Reply by zinginging** [reply]
I use master branch and [docker compose](https://superset.apache.org/docs/installation/installing-superset-using-docker-compose/)  ,works fine finally. Maybe  some initialization steps are missing when using`docker run -d xx` command .

**Reply by JanDoerrenhaus** [reply]
I have exactly the same problem when doing the aforementioned steps using the latest Superset.

I had to add the install for the specific `holidays` version, since without it, setting up the admin account failed with error message: `The 'holidays<0.26,>=0.25' distribution was not found and is required by apache-superset`.

```
FROM apache/superset:latest
USER root
RUN pip install 'shillelagh[all]'
RUN pip install 'holidays==0.25'
USER superset
```

Then I am doing the same setup steps as mentioned by the OP, creating a Shillelagh database, but then, no matter which URL I try (even nonexistent ones), I always just get `shillelagh error: unable to open database file`.

This is a blocker for the PoC that I am supposed to do. Any help would be greatly appreciated.

**Reply by betodealmeida** [reply]
@JanDoerrenhaus I'll take a look today.

**Reply by JanDoerrenhaus** [reply]
@betodealmeida Any news?

**Reply by JanDoerrenhaus** [reply]
I am really running out of time here, sorry for being a pest. I don't know what happened to @betodealmeida, can anyone help me at all?

**Reply by betodealmeida** [reply]
@JanDoerrenhaus sorry, I didn't have time last week and this week I'm on vacation.

I think the problem is caused because Shillelagh uses SQLite, and even for an in-memory database it still needs access to a directory like `/tmp`. I suspect the Docker configuration is preventing the file from being created.

Can you try setting the env variable `SQLITE_TMPDIR` inside Docker and have it point to a directory that exists? Maybe `/app/superset/tmp`?

**Reply by JanDoerrenhaus** [reply]
@betodealmeida Very sorry to bother you during your vacation.

I adapted my Dockerfile like so:
```
FROM apache/superset:latest
USER root
RUN pip install 'shillelagh[all]'
RUN pip install 'holidays==0.25'
RUN mkdir /app/superset/tmp
RUN chown superset /app/superset/tmp
USER superset
ENV SQLITE_TMPDIR=/app/superset/tmp
```

Unfortunately, that did not change anything about the error.
```
shillelagh error: unable to open database file
```

**Reply by vigodeltoro** [reply]
Hi over there. 
I can also verify @JanDoerrenhaus issue. I'm also needing shillelagh running for a PoC. 

Superset 3.1.3
shillelagh 1.2.10
holidays 0.25

Running Superset in Docker/Kubernetes

With adding an ENV variable I'm running in the same issue. 
I also tested the html function 

SELECT *
FROM "https://en.wikipedia.org/wiki/List_of_countries_and_dependencies_by_population"
WHERE "UN Region" = 'Oceania'
LIMIT 5

Same issue.. 

Any help would be very appreciated :).. 
best and thanks

**Reply by vigodeltoro** [reply]
Hi, 

I can confirm that shillelagh is functional without docker. I setup a vm and installed Superset (3.1.3) via PyPi. Shillelagh 1.2.19 was installed automatically. I added `pip3 install shillelagh[all].` 

And tried :
`SELECT
*
FROM "https://data.cdc.gov/resource/unsk-b7fc.json"
WHERE location = 'US'
ORDER BY date DESC
LIMIT 10;`

Setting the location of the shillelagh sqlite cache file via `sqlite3_temp_directory` env variable is also working.

BUT: Everything with SQLITE dev Metadatabase.

If I try to transfer that to productional Superset  with Postgres Metadatabase I got the error ( with and without env variable for setting location of cache sqlite file):

`shillelagh error: unable to open database file`

So, any help would be appreciated

best and thanks

**Reply by vigodeltoro** [reply]
Was able to reproduce the issue in docker and Postgres meta database. If supererset is not running as root than it can't write the file socrata.cache.sqlite to /app (which seems to be the default). If it's running as root it is working. So it's a permission denied error. 

I wasn't able to configure a working env variable for the sqlite tmp files (sqlite3_temp_directory, SQLITE_TMPDIR) yet.

**Comment by vigodeltoro** [comment]
Okay.. I was able to fix it.. but only with a modified superset docker image. 
@JanDoerrenhaus what you have to do is in your docker file:

FROM apache/superset:latest
USER root
RUN pip install 'shillelagh[all]'
RUN pip install 'holidays==0.25'
RUN chown superset:superset /app
USER superset

best regards, hope that helps you..
